File: lua/modules/editor/vscode.lua
-- sysinit.nvim.doc-url="https://raw.githubusercontent.com/vscode-neovim/vscode-neovim/main/README.md"
local M = {}

-- Map of Neovim commands (without <cmd> and <cr>) to VSCode action names
M.cmd_map = {
  w      = 'workbench.action.files.save',
  wa     = 'workbench.action.files.saveAll',
  q      = 'workbench.action.closeActiveEditor',
  qa     = 'workbench.action.quit',
  enew   = 'workbench.action.files.newUntitledFile',
  bdelete= 'workbench.action.closeActiveEditor',
  bn     = 'workbench.action.nextEditor',
  bp     = 'workbench.action.previousEditor',
  split  = 'workbench.action.splitEditorDown',
  vsplit = 'workbench.action.splitEditorRight',
}

-- VSCode plugin spec for Lazy.nvim
M.plugins = {
  {
    "vscode-neovim/vscode-neovim",
    lazy = false,
    cond = function()
      return vim.g.vscode == true
    end,
  },
}

--- Mirror <leader> keymaps from Neovim to VSCode
function M.setup_layer()
  if not vim.g.vscode then return end
  local api = require('vscode')
  local leader = vim.g.mapleader or ' '
  local summary = {}

  for _, m in ipairs(vim.api.nvim_get_keymap('n')) do
    if m.lhs:match('^' .. leader) then
      local lhs = m.lhs
      local rhs = m.rhs
      local opts = { noremap = m.noremap == 1, silent = m.silent == 1 }
      local cmd = rhs:match('^<cmd>(.+)<cr>$')
      if cmd then
        local action = M.cmd_map[cmd]
        if action then
          vim.keymap.set('n', lhs, function() api.action(action) end, opts)
          table.insert(summary, lhs .. ' -> ' .. action)
        else
          vim.keymap.set('n', lhs, '<cmd>' .. cmd .. '<cr>', opts)
          table.insert(summary, lhs .. ' -> ' .. cmd)
        end
      end
    end
  end

  -- Command to list mirrored keymaps
  vim.api.nvim_create_user_command('VsKeymaps', function()
    for _, line in ipairs(summary) do
      vim.api.nvim_echo({{line, 'Normal'}}, false, {})
    end
  end, {})
end

return M

File: pkg/modules/vscode.go
package modules

import (
   "fmt"
   "regexp"
   "strings"

   "github.com/roshbhatia/nix-neovim-config/pkg/core"
)

// VscodeModule mirrors Neovim keybindings in VSCode via vscode-neovim
type VscodeModule struct{}

// NewVscodeModule constructs a new VscodeModule.
func NewVscodeModule() *VscodeModule {
   return &VscodeModule{}
}

// Name returns the module name.
func (m *VscodeModule) Name() string {
   return "vscode"
}

// Setup registers VSCode keymaps mirroring Neovim leader mappings.
// It only runs when 'vim.g.vscode' is true.
func (m *VscodeModule) Setup(ctx *core.Context) error {
   // Skip setup when running manifest generation (no Nvim client)
   if ctx.Nvim == nil {
       return nil
   }
   // Detect VSCode environment via vim.g.vscode
   var enabled bool
   if err := ctx.Nvim.Eval("vim.g.vscode == true", &enabled); err != nil || !enabled {
       return nil
   }

   // Mapping of Neovim commands to VSCode actions
   cmdMap := map[string]string{
       "w":      "workbench.action.files.save",
       "wa":     "workbench.action.files.saveAll",
       "q":      "workbench.action.closeActiveEditor",
       "qa":     "workbench.action.quit",
       "enew":   "workbench.action.files.newUntitledFile",
       "bdelete": "workbench.action.closeActiveEditor",
       "bn":     "workbench.action.nextEditor",
       "bp":     "workbench.action.previousEditor",
       "split":  "workbench.action.splitEditorDown",
       "vsplit": "workbench.action.splitEditorRight",
   }

   // Determine leader key (default to space)
   leader := " "
   var ldr string
   if err := ctx.Nvim.Eval("vim.g.mapleader", &ldr); err == nil && ldr != "" {
       leader = ldr
   }

   // Retrieve all normal mode keymaps
   var keymaps []map[string]interface{}
   if err := ctx.Nvim.Eval("vim.api.nvim_get_keymap('n')", &keymaps); err != nil {
       return err
   }

   // Regex to extract command from <cmd>...<cr>
   re := regexp.MustCompile(`^<cmd>(.+)<cr>$`)

   // Mirror each leader mapping
   for _, km := range keymaps {
       lhsVal, ok := km["lhs"].(string)
       if !ok || !strings.HasPrefix(lhsVal, leader) {
           continue
       }
       rhsVal, ok := km["rhs"].(string)
       if !ok {
           continue
       }
       // Extract options
       opts := map[string]bool{}
       if v, ok := km["noremap"].(int64); ok {
           opts["noremap"] = v != 0
       }
       if v, ok := km["silent"].(int64); ok {
           opts["silent"] = v != 0
       }
       // Match <cmd>...<cr>
       if m := re.FindStringSubmatch(rhsVal); len(m) == 2 {
           cmdText := m[1]
           if action, found := cmdMap[cmdText]; found {
               // Map to VSCode action via Lua call
               rhs := fmt.Sprintf("<cmd>lua require('vscode').action(%q)<cr>", action)
               if err := ctx.Map("n", lhsVal, rhs, opts); err != nil {
                   return err
               }
           } else {
               // Fallback: original command mapping
               if err := ctx.Map("n", lhsVal, rhsVal, opts); err != nil {
                   return err
               }
           }
       }
   }
   return nil
}